staging/src/k8s.io/component-base/logs/api/v1/kube_features.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"k8s.io/component-base/featuregate"
+)
+
+const (
+	// owner: @pohly
+	// alpha: v1.24
+	// beta: see LoggingBetaOptions
+	//
+	// Allow fine-tuning of experimental, alpha-quality logging options.
+	//
+	// Per https://groups.google.com/g/kubernetes-sig-architecture/c/Nxsc7pfe5rw/m/vF2djJh0BAAJ
+	// we want to avoid a proliferation of feature gates. This feature gate:
+	// - will guard *a group* of logging options whose quality level is alpha.
+	// - will never graduate to beta or stable.
+	LoggingAlphaOptions featuregate.Feature = "LoggingAlphaOptions"
+
+	// owner: @pohly
+	// alpha: see LoggingAlphaOptions
+	// beta: v1.24
+	//
+	// Allow fine-tuning of experimental, beta-quality logging options.
+	//
+	// Per https://groups.google.com/g/kubernetes-sig-architecture/c/Nxsc7pfe5rw/m/vF2djJh0BAAJ
+	// we want to avoid a proliferation of feature gates. This feature gate:
+	// - will guard *a group* of logging options whose quality level is beta.
+	// - is thus *introduced* as beta
+	// - will never graduate to stable.
+	LoggingBetaOptions featuregate.Feature = "LoggingBetaOptions"
+)
+
+func init() {
+	if err := featuregate.DefaultMutableFeatureGate.Add(features); err != nil {
+		panic(err)
+	}
+}
+
+var features = map[featuregate.Feature]featuregate.FeatureSpec{
+	LoggingAlphaOptions: {Default: false, PreRelease: featuregate.Alpha},
+	LoggingBetaOptions:  {Default: true, PreRelease: featuregate.Beta},
```

Comment:
I'm not sure if I like this... this implies that we will have perma alpha and beta feature gates and it will be difficult to track which logging features are behind which gates, and which ones have graduated.

Is it possible to model this after how we normally do feature gates, where each feature has its own graduation cycle?

Reply:
This was discussed before in the context of CPUManagerPolicyAlphaOptions and CPUManagerPolicyBetaOptions, see https://groups.google.com/g/kubernetes-sig-architecture/c/Nxsc7pfe5rw

The key arguments why combining multiple features is better in cases like this are:
- Each of the features gets enabled via a flag or configuration option, not via `--feature-gates`, i.e. fine-grained control over features is still possible.
- It avoids overwhelming the user with too many features listed for `--feature-gates`.
- It avoids breaking user's command line invocations when a feature graduates: `kube-scheduler --feature-gate LogJSONSplitStream=true --log-json-split-stream` would become invalid at some point, but `kube-scheduler --feature-gate LoggingAlphaOptions=true --log-json-split-stream` remains valid.

Regarding "it will be difficult to track which logging features are behind which gates": that is addressed by the alpha/beta comments in LoggingConfiguration and the command line flag descriptions. The feature name could be added to LoggingConfiguration if we want to be really precise. For the command line it might become too long.